### PR TITLE
[WIP] Check JavaScript files using eslint.

### PR DESCRIPTION
**Pull request by fancycode**
@MorrisJobke: as suggested in https://github.com/nextcloud/spreed/pull/48 I added the eslint checker here, too.

Please take a look at the Done output for the reported errors in the current code. If these are fine for you to be checked/reported, I will update this PR to add commits that fix the errors.

### CODE REVIEW

**Comment by MorrisJobke** [comment]
That was fast - I suspect a ton of issues here. But maybe we could keep that as a good starting point and iterate on this :) Thanks a lot @fancycode

**Comment by fancycode** [comment]
For reference, you can see the errors here: https://drone.nextcloud.com/nextcloud/server/752/22

Imho all found issues should be fixed (and not silenced) to keep the code clean. Bonus points for fixing and enabling the rules `no-mixed-spaces-and-tabs` and `no-undef` but especially the last one is probably too much work given that everything is global ;-)

File: .drone.yml [hunk 1]
```diff
@@ -9,6 +9,13 @@ pipeline:
     when:
       matrix:
         TESTS: jsunit
+  eslint:
+    image: nextcloudci/eslint:eslint-1
+    commands:
+      - ./run-eslint.sh
```
Comment (LukasReschke) [inline .drone.yml]: Can you move that to the build folder?

**Comment by MorrisJobke** [comment]
Remaining issues:

```
settings/js/admin.js
  112:4   error  'savePublicShareDisclaimerText' was used before it was defined  no-use-before-define
settings/js/users/filter.js
  58:24  error  Expected '===' and instead saw '=='  eqeqeq
settings/js/users/users.js
  183:11  error  Expected a conditional expression and instead saw an assignment                                             no-cond-assign
  296:3   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  311:3   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  496:8   error  Redundant double negation                                                                                   no-extra-boolean-cast
  704:35  error  'displayName' was used before it was defined                                                                no-use-before-define
  845:7   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
settings/js/personal.js
   88:6   error  'updateAvatar' was used before it was defined           no-use-before-define
  144:14  error  'saveCoords' was used before it was defined             no-use-before-define
  145:14  error  'saveCoords' was used before it was defined             no-use-before-define
  155:2   error  'cleanCropper' was used before it was defined           no-use-before-define
  164:58  error  'avatarResponseHandler' was used before it was defined  no-use-before-define
settings/js/authtoken_view.js
  152:9   error  Expected a conditional expression and instead saw an assignment  no-cond-assign
core/search/js/search.js
  111:24  error  'getCurrentApp' was used before it was defined  no-use-before-define
  131:7   error  '$status' was used before it was defined        no-use-before-define
  132:7   error  '$status' was used before it was defined        no-use-before-define
  138:9   error  'showResults' was used before it was defined    no-use-before-define
  140:9   error  'addResults' was used before it was defined     no-use-before-define
  181:5   error  'addResults' was used before it was defined     no-use-before-define
core/js/placeholders.js
  152:23  error  Empty block statement  no-empty
core/js/multiselect.js
   63:23  error  Expected '===' and instead saw '=='  eqeqeq
  232:27  error  Expected '===' and instead saw '=='  eqeqeq
  232:54  error  Expected '===' and instead saw '=='  eqeqeq
core/js/oc-dialogs.js
  217:17  error  Empty block statement                             no-empty
  340:17  error  'crop' was used before it was defined             no-use-before-define
  375:4   error  'resampleHermite' was used before it was defined  no-use-before-define
  430:10  error  Expected { after 'if' condition                   curly
  802:7   error  'dir' is defined but never used                   no-unused-vars
core/js/compatibility.js
  141:10  error  'outerHTML' is defined but never used  no-unused-vars
core/js/eventsource.js
  48:3  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/js.js
    11:5   error  'oc_debug' is defined but never used                no-unused-vars
    45:10  error  'escapeHTML' is defined but never used              no-unused-vars
    53:19  error  Irregular whitespace not allowed                    no-irregular-whitespace
    56:10  error  'fileDownloadPath' is defined but never used        no-unused-vars
    57:9   error  'OC' was used before it was defined                 no-use-before-define
   152:8   error  'vars' is already defined                           no-redeclare
   169:34  error  Expected '===' and instead saw '=='                 eqeqeq
   188:13  error  Expected '!==' and instead saw '!='                 eqeqeq
   205:13  error  Expected '===' and instead saw '=='                 eqeqeq
   205:34  error  Expected '===' and instead saw '=='                 eqeqeq
   205:51  error  Expected '===' and instead saw '=='                 eqeqeq
   205:72  error  Expected '===' and instead saw '=='                 eqeqeq
   327:23  error  Expected '===' and instead saw '=='                 eqeqeq
   501:4   error  'value' is defined but never used                   no-unused-vars
   564:28  error  Irregular whitespace not allowed                    no-irregular-whitespace
   590:8   error  'jqxhr' is defined but never used                   no-unused-vars
  1164:7   error  'self' is defined but never used                    no-unused-vars
  1339:28  error  'item' is defined but never used                    no-unused-vars
  1378:10  error  'object' is defined but never used                  no-unused-vars
  1600:29  error  Expected '===' and instead saw '=='                 eqeqeq
  1651:78  error  'getScrollBarWidth' was used before it was defined  no-use-before-define
  1656:47  error  'getScrollBarWidth' was used before it was defined  no-use-before-define
  1731:10  error  'formatDate' is defined but never used              no-unused-vars
  1742:10  error  'getURLParameter' is defined but never used         no-unused-vars
  1745:22  error  Unexpected comma in middle of array                 no-sparse-arrays
  1753:10  error  'relative_modified_date' is defined but never used  no-unused-vars
  1816:23  error  '$el' is defined but never used                     no-unused-vars
  1827:27  error  '$el' is defined but never used                     no-unused-vars
  1901:38  error  'code' is defined but never used                    no-unused-vars
  1934:14  error  Expected '===' and instead saw '=='                 eqeqeq
  1934:31  error  Expected '===' and instead saw '=='                 eqeqeq
core/js/files/iedavclient.js
  40:4  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/files/client.js
  464:4  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/share.js
  108:3   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  131:19  error  Expected '!==' and instead saw '!='                                                                         eqeqeq
  162:7   error  'image' is defined but never used                                                                           no-unused-vars
  166:15  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  167:37  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  179:16  error  Expected '!==' and instead saw '!='                                                                         eqeqeq
  179:38  error  Expected '!==' and instead saw '!='                                                                         eqeqeq
  214:7   error  'dir' is defined but never used                                                                             no-unused-vars
  333:8   error  'checkShares' is already defined                                                                            no-redeclare
core/js/shareitemmodel.js
  150:8  error  'shareType' is defined but never used  no-unused-vars
core/js/sharedialogview.js
  116:4  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/maintenance-check.js
  2:20  error  'checkStatus' was used before it was defined  no-use-before-define
core/js/singleselect.js
  57:31  error  Expected '===' and instead saw '=='  eqeqeq
  75:27  error  Expected '===' and instead saw '=='  eqeqeq
core/js/lostpassword.js
   54:29  error  'msg' is defined but never used  no-unused-vars
  118:22  error  'msg' is defined but never used  no-unused-vars
core/js/placeholder.js
  100:3  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/tags.js
    2:22  error  'cb' is defined but never used     no-unused-vars
   26:59  error  'event' is defined but never used  no-unused-vars
   29:41  error  'event' is defined but never used  no-unused-vars
   63:28  error  'ui' is defined but never used     no-unused-vars
  176:4   error  'self' is defined but never used   no-unused-vars
  362:18  error  'tag' is defined but never used    no-unused-vars
core/js/jquery-showpassword.js
  113:44  error  'e' is defined but never used  no-unused-vars
core/js/oc-backbone-webdav.js
   93:4   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  104:15  error  'parseIdFromLocation' was used before it was defined                                                        no-use-before-define
  133:3   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
core/js/octemplate.js
  97:7  error  'vars' is already defined  no-redeclare
apps/federation/js/settings-admin.js
  67:22  error  'response' is defined but never used  no-unused-vars
apps/twofactor_backupcodes/js/settingsview.js
  1:11  error  'Backbone' is defined but never used  no-unused-vars
apps/theming/js/settings-admin.js
   28:3   error  'hideUndoButton' was used before it was defined  no-use-before-define
   32:2   error  'preview' was used before it was defined         no-use-before-define
  146:8   error  'value' is already defined                       no-redeclare
  159:23  error  'response' is defined but never used             no-unused-vars
  177:23  error  'response' is defined but never used             no-unused-vars
  195:17  error  Expected '===' and instead saw '=='              eqeqeq
  205:17  error  Expected '===' and instead saw '=='              eqeqeq
  215:17  error  Expected '===' and instead saw '=='              eqeqeq
  220:39  error  'e' is defined but never used                    no-unused-vars
  224:35  error  'e' is defined but never used                    no-unused-vars
apps/files_sharing/js/public.js
  159:60  error  'isDir' is defined but never used  no-unused-vars
apps/files_sharing/js/sharedfilelist.js
  181:9  error  Redundant double negation         no-extra-boolean-cast
  227:8  error  'self' is defined but never used  no-unused-vars
apps/workflowengine/js/admin.js
  176:45  error  'options' is defined but never used     no-unused-vars
  184:43  error  'options' is defined but never used     no-unused-vars
  230:66  error  'check' was used before it was defined  no-use-before-define
apps/files_trashbin/js/app.js
  21:23  error  '$el' is defined but never used  no-unused-vars
apps/systemtags/js/systemtagsinfoview.js
  42:24  error  'options' is defined but never used  no-unused-vars
apps/files/js/favoritesfilelist.js
  39:30  error  'options' is defined but never used  no-unused-vars
apps/files/js/file-upload.js
   29:9   error  'supportFileAPI' was used before it was defined                   no-use-before-define
   29:29  error  'supportAjaxUploadProgressEvents' was used before it was defined  no-use-before-define
   29:66  error  'supportFormData' was used before it was defined                  no-use-before-define
  120:20  error  'data' is defined but never used                                  no-unused-vars
  412:35  error  'selection' is defined but never used                             no-unused-vars
  415:38  error  'selection' is defined but never used                             no-unused-vars
  418:37  error  'selection' is defined but never used                             no-unused-vars
  523:11  error  'original' is already defined                                     no-redeclare
  524:11  error  'replacement' is already defined                                  no-redeclare
  730:41  error  'e' is defined but never used                                     no-unused-vars
apps/files/js/breadcrumb.js
  89:31  error  'index' is defined but never used  no-unused-vars
apps/files/js/detailtabview.js
  107:24  error  'fileInfo' is defined but never used  no-unused-vars
apps/files/js/filelist.js
   385:40  error  'context' is defined but never used          no-unused-vars
   548:39  error  'showDetailsView' is defined but never used  no-unused-vars
   774:23  error  'e' is defined but never used                no-unused-vars
  1041:34  error  'options' is defined but never used          no-unused-vars
  2374:19  error  'query' is defined but never used            no-unused-vars
  2447:22  error  'filter' is defined but never used           no-unused-vars
apps/files/js/newfilemenu.js
  190:22  error  Expected '===' and instead saw '=='  eqeqeq
apps/files/js/jquery.fileupload.js
    15:1   error  Unnecessary semicolon                       no-extra-semi
  1102:21  error  'dirReader' was used before it was defined  no-use-before-define
  1104:44  error  'entries' was used before it was defined    no-use-before-define
  1106:29  error  'entries' was used before it was defined    no-use-before-define
  1106:39  error  'entries' was used before it was defined    no-use-before-define
apps/files/js/files.js
   11:11  error  'getURLParameter' is defined but never used    no-unused-vars
   86:30  error  Expected '===' and instead saw '=='            eqeqeq
  389:25  error  'ui' is defined but never used                 no-unused-vars
  396:24  error  'ui' is defined but never used                 no-unused-vars
  407:24  error  'ui' is defined but never used                 no-unused-vars
  434:5   error  'folderDropOptions' is defined but never used  no-unused-vars
  464:10  error  'fileDownloadPath' is defined but never used   no-unused-vars
apps/files/js/keyboardshortcuts.js
  102:35  error  'index' is defined but never used   no-unused-vars
  111:35  error  'index' is defined but never used   no-unused-vars
  119:11  error  'rename' is defined but never used  no-unused-vars
  120:35  error  'index' is defined but never used   no-unused-vars
apps/files/js/jquery-visibility.js
  12:1  error  Unnecessary semicolon  no-extra-semi
apps/files/js/tagsplugin.js
  95:10  error  'fileInfo' is defined but never used  no-unused-vars
apps/files/js/app.js
  127:35  error  'newAction' is defined but never used  no-unused-vars
apps/files/js/search.js
   44:13  error  'updateLegacyMimetype' is defined but never used  no-unused-vars
  134:52  error  'event' is defined but never used                 no-unused-vars
  144:50  error  'event' is defined but never used                 no-unused-vars
apps/files/js/filesummary.js
  106:9  error  Redundant double negation  no-extra-boolean-cast
  132:9  error  Redundant double negation  no-extra-boolean-cast
apps/user_ldap/js/wizard/view.js
  285:4   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  420:19  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  435:19  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  449:20  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  453:20  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  457:20  error  Expected '===' and instead saw '=='                                                                         eqeqeq
  461:20  error  Expected '===' and instead saw '=='                                                                         eqeqeq
apps/user_ldap/js/wizard/wizardTabElementary.js
  187:6  error  Unreachable code  no-unreachable
  195:6  error  Unreachable code  no-unreachable
apps/user_ldap/js/wizard/wizardObject.js
  28:3  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
apps/user_ldap/js/wizard/wizardDetectorGeneric.js
   87:24  error  'configID' is defined but never used                                                                        no-unused-vars
  102:5   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
apps/user_ldap/js/wizard/wizardTabGeneric.js
   84:43  error  'key' is defined but never used                                                                             no-unused-vars
  231:4   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  281:53  error  Expected '===' and instead saw '=='                                                                         eqeqeq
apps/encryption/js/settings-admin.js
  52:21  error  Expected '===' and instead saw '=='  eqeqeq
apps/files_external/js/settings.js
  658:7  error  'self' is defined but never used  no-unused-vars
apps/files_external/js/public_key.js
   7:5  error  'setupTableRow' was used before it was defined  no-use-before-define
  11:7  error  'generateKeys' was used before it was defined   no-use-before-define
  21:3  error  'generateKeys' was used before it was defined   no-use-before-define
apps/files_external/js/oauth1.js
  19:29  error  Expected '===' and instead saw '=='  eqeqeq
  24:18  error  Expected '!==' and instead saw '!='  eqeqeq
  24:38  error  Expected '!==' and instead saw '!='  eqeqeq
  27:15  error  Expected '!==' and instead saw '!='  eqeqeq
  27:82  error  Expected '===' and instead saw '=='  eqeqeq
  33:37  error  Expected '===' and instead saw '=='  eqeqeq
  58:15  error  Expected '!==' and instead saw '!='  eqeqeq
  58:35  error  Expected '!==' and instead saw '!='  eqeqeq
  63:33  error  Expected '===' and instead saw '=='  eqeqeq
apps/files_external/js/mountsfilelist.js
  36:29  error  'options' is defined but never used  no-unused-vars
apps/files_external/js/dropbox.js
  12:77  error  'onCompletion' is defined but never used  no-unused-vars
apps/files_external/js/oauth2.js
  19:29  error  Expected '===' and instead saw '=='  eqeqeq
  25:20  error  Expected '!==' and instead saw '!='  eqeqeq
  25:43  error  Expected '!==' and instead saw '!='  eqeqeq
  43:39  error  Expected '===' and instead saw '=='  eqeqeq
  71:17  error  Expected '!==' and instead saw '!='  eqeqeq
  71:40  error  Expected '!==' and instead saw '!='  eqeqeq
  80:34  error  Expected '===' and instead saw '=='  eqeqeq
  83:72  error  'status' is defined but never used   no-unused-vars
apps/files_external/js/statusmanager.js
  523:31  error  'index' is defined but never used  no-unused-vars
  601:25  error  'cobj' is defined but never used   no-unused-vars
apps/files_external/js/gdrive.js
  3:23  error  '$tr' is defined but never used           no-unused-vars
  8:77  error  'onCompletion' is defined but never used  no-unused-vars
```

**Comment by rullzer** [comment]
It seems we have a lot of `is defined but never used` warnings. Sometime this is just from function signatures that chaning is not to easy.

we could create an eslint rule to variables that start with `ignore`. Everybody up for that?

**Comment by LukasReschke** [comment]
Closing as there hasn't been any progress on this for more than 1 month and there are merge commits. Please reopen if you want to continue to work on this.

**Pull request closed without merging**